File: plugin/jvm/src/main/java/com/twosigma/beaker/chart/categoryplot/CategoryPlot.java
/*
 *  Copyright 2014 TWO SIGMA OPEN SOURCE, LLC
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.twosigma.beaker.chart.categoryplot;

import com.twosigma.beaker.AbstractChart;
import com.twosigma.beaker.chart.categoryplot.plotitem.CategoryGraphics;

import java.util.ArrayList;
import java.util.List;

public class CategoryPlot extends AbstractChart {
  private final List<CategoryGraphics> categoryGraphics = new ArrayList<>();
  private List<String> categoryNames = new ArrayList<>();


  public CategoryPlot leftShift(CategoryGraphics graphics) {
    return add(graphics);
  }

  public List<CategoryGraphics> getGraphics() {
    return this.categoryGraphics;
  }

  public CategoryPlot add(CategoryGraphics graphics) {
    this.categoryGraphics.add(graphics);
    return this;
  }

  public CategoryPlot add(List items) {
    for (Object o : items) {
      if (o instanceof CategoryGraphics) {
        add((CategoryGraphics) o);
      } else {
        super.add(items);
      }
    }
    return this;
  }

  public List<String> getCategoryNames() {
    return categoryNames;
  }

  public CategoryPlot setCategoryNames(List<String> categoryNames) {
    this.categoryNames = categoryNames;
    return this;
  }
}

File: kernel/scala/src/main/scala/com/twosigma/beakerx/scala/chart/categoryplot/CategoryPlot.scala
/*
 * Copyright 2018 TWO SIGMA OPEN SOURCE, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.twosigma.beakerx.scala.chart.categoryplot

import com.twosigma.beakerx.chart.categoryplot.plotitem.CategoryGraphics
import com.twosigma.beakerx.chart.xychart.plotitem.PlotOrientationType
import com.twosigma.beakerx.scala.chart.AbstractChartProperties

import scala.collection.JavaConverters._

class CategoryPlot extends com.twosigma.beakerx.chart.categoryplot.CategoryPlot with CategoryPlotProperties

trait CategoryPlotProperties extends AbstractChartProperties {
  this: com.twosigma.beakerx.chart.categoryplot.CategoryPlot =>

  def categoryMargin: Double = getCategoryMargin
  def categoryMargin_=(margin: Double): Unit = setCategoryMargin(margin)

  def categoryNames: Seq[String] = getCategoryNames.asScala
  def categoryNames_=(names: Seq[String]): Unit = setCategoryNames(names.asJava)

  def categoryNamesLabelAngle: Double = getCategoryNamesLabelAngle
  def categoryNamesLabelAngle_=(angle: Double): Unit = setCategoryNamesLabelAngle(angle)

  def orientation: PlotOrientationType = getOrientation
  def orientation_=(orientation: PlotOrientationType): Unit = setOrientation(orientation)

  def +=(cg: CategoryGraphics): Unit = add(cg)
}
